src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/ICompileLibrariesProvider.cs

```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    /// <summary>
+    /// Exposes <see cref="CompilationLibrary"/> instances from an <see cref="ApplicationPart"/>.
+    /// </summary>
+    public interface ICompileLibrariesProvider
+    {
+        /// <summary>
+        /// Gets the sequence of <see cref="CompilationLibrary"/> instances.
+        /// </summary>
+        IReadOnlyList<CompilationLibrary> GetCompilationLibraries();
```

Comment:
> Why wouldn't they just add a metadata reference?

You can ask this question about 100% of the application parts feature, but we still lots of requests for something like this and we still ended up building it. You can solve all of the problems that application model solves by just referencing everything from one project, but users have come up with plenty of cases where they don't want to do that.

As @pranavkm said this is here so you can author a custom part and have control over which assemblies are included as references. Providing interfaces for the various capabilities is part of the design of parts. 

Why would you do this? Well I suppose you could use it to work around the bug we have where a .csproj isn't part of the dependency graph :trollface: